src/Microsoft.Health.Dicom.Core/Features/Retrieve/RetrieveRenderedService.cs

```diff
@@ -0,0 +1,90 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using FellowOakDicom;
+using FellowOakDicom.Imaging;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Dicom.Core.Exceptions;
+using Microsoft.Health.Dicom.Core.Extensions;
+using Microsoft.Health.Dicom.Core.Features.Common;
+using Microsoft.Health.Dicom.Core.Features.Context;
+using Microsoft.Health.Dicom.Core.Features.Model;
+using Microsoft.Health.Dicom.Core.Messages;
+using Microsoft.Health.Dicom.Core.Messages.Retrieve;
+using SixLabors.ImageSharp;
+
+namespace Microsoft.Health.Dicom.Core.Features.Retrieve;
+public class RetrieveRenderedService : IRetrieveRenderedService
+{
+    private readonly IFileStore _blobDataStore;
+    private readonly IInstanceStore _instanceStore;
+    private readonly IDicomRequestContextAccessor _dicomRequestContextAccessor;
+    private readonly ILogger<RetrieveResourceService> _logger;
+
+    public RetrieveRenderedService(
+        IInstanceStore instanceStore,
+        IFileStore blobDataStore,
+        IDicomRequestContextAccessor dicomRequestContextAccessor,
+        ILogger<RetrieveResourceService> logger)
+    {
+        EnsureArg.IsNotNull(instanceStore, nameof(instanceStore));
+        EnsureArg.IsNotNull(blobDataStore, nameof(blobDataStore));
+        EnsureArg.IsNotNull(dicomRequestContextAccessor, nameof(dicomRequestContextAccessor));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        _instanceStore = instanceStore;
+        _blobDataStore = blobDataStore;
+        _dicomRequestContextAccessor = dicomRequestContextAccessor;
+        _logger = logger;
+    }
+
+    public async Task<RetrieveRenderedResponse> RetrieveRenderedImageAsync(RetrieveRenderedRequest request, CancellationToken cancellationToken = default)
+    {
+        EnsureArg.IsNotNull(request, nameof(request));
+        var partitionKey = _dicomRequestContextAccessor.RequestContext.GetPartitionKey();
+
+        try
+        {
+            // this call throws NotFound when zero instance found
+            IEnumerable<InstanceMetadata> retrieveInstances = await _instanceStore.GetInstancesWithProperties(
+                ResourceType.Instance, partitionKey, request.StudyInstanceUid, request.SeriesInstanceUid, request.SopInstanceUid, cancellationToken);
+
+            InstanceMetadata instance = retrieveInstances.First();
+
+            _dicomRequestContextAccessor.RequestContext.PartCount = retrieveInstances.Count();
+
+            Stream stream = await _blobDataStore.GetFileAsync(instance.VersionedInstanceIdentifier, cancellationToken);
+
+            DicomFile dicomFile = await DicomFile.OpenAsync(stream, FileReadOption.ReadLargeOnDemand);
+
+            var dicomImage = new DicomImage(dicomFile.Dataset);
+            using var img = dicomImage.RenderImage(request.FrameNumber);
+            using var sharpImage = img.AsSharpImage();
```

Comment:
Do you think it will be useful to know how much time we spent on this particular method?